### PR TITLE
Distributed Anomaly Detection

### PR DESCRIPTION
**Pull request by npawar**
This pull request contains the distribued anomaly detection code for scheduler and worker

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyApplication.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    bootstrap.addBundle(hibernateBundle);
+
+    bootstrap.addBundle(new AssetsBundle("/assets/", "/", "index.html"));
+  }
+
+  @Override
+  public void run(final ThirdEyeAnomalyConfiguration config, final Environment environment)
+      throws Exception {
+
+    System.out.println("Starting............");
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyApplication.java]: Use logger

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyApplication.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+
+    environment.lifecycle().manage(new Managed() {
+      @Override
+      public void start() throws Exception {
+        new HibernateSessionWrapper<Void>(hibernateBundle.getSessionFactory())
+            .execute(new Callable<Void>() {
+          @Override
+          public Void call() throws Exception {
+            if (config.isWorker()) {
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyApplication.java]: init driver/scheduler only if required.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/JobRunner.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+import com.linkedin.thirdeye.detector.db.AnomalyFunctionSpecDAO;
+import com.linkedin.thirdeye.detector.db.AnomalyJobSpecDAO;
+import com.linkedin.thirdeye.detector.db.AnomalyTaskSpecDAO;
+
+public class JobRunner implements Job {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JobRunner.class);
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+  static {
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/JobRunner.java]: remove

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/JobRunner.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    }
+    thirdEyeJobContext.setWindowStart(windowStart);
+    thirdEyeJobContext.setWindowEnd(windowEnd);
+
+    // write to anomaly_jobs
+    Long jobExecutionId = createAnomlyJob(jobName);
+    thirdEyeJobContext.setJobExecutionId(jobExecutionId);
+
+    // write to anomaly_tasks
+    List<Long> taskIds = createAnomlyTasks(anomalyFunctionSpec);
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/JobRunner.java]: s/Anomly/Anomaly/

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/api/AnomalyJobSpec.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+import com.google.common.base.MoreObjects;
+import com.linkedin.thirdeye.anomaly.JobRunner;
+import com.linkedin.thirdeye.anomaly.JobRunner.JobStatus;
+
+@Entity
+@Table(name = "anomaly_jobs")
+@NamedQueries({
+    @NamedQuery(name = "com.linkedin.thirdeye.anomaly.AnomalyJobSpec#findAll", query = "SELECT af FROM AnomalyJobSpec af")
+})
+public class AnomalyJobSpec {
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/api/AnomalyJobSpec.java]: javadoc explainig purpose and difference from *Task

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/TaskContext.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import com.linkedin.thirdeye.detector.db.AnomalyFunctionRelationDAO;
+import com.linkedin.thirdeye.detector.db.AnomalyResultDAO;
+
+public class TaskContext {
+
+  PinotThirdEyeClient thirdEyeClient;
+  private AnomalyResultDAO resultDAO;
+  private AnomalyFunctionRelationDAO relationDAO;
+  private SessionFactory sessionFactory;
+  public PinotThirdEyeClient getThirdEyeClient() {
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/TaskContext.java]: Remove if unused, or change to ThirdEyeClient

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/AnomalyTaskSpecDAO.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            .setParameter("status", status));
+  }
+
+  // also update the worker id that is picking up the task
+  public boolean updateStatus(Long taskId, JobStatus oldStatus, JobStatus newStatus) {
+    try {
+      int numRowsUpdated = namedQuery("com.linkedin.thirdeye.anomaly.AnomalyTaskSpec#updateStatus")
+          .setParameter("taskId", taskId).setParameter("oldStatus", oldStatus).setParameter("newStatus", newStatus)
+          .executeUpdate();
+      System.out.println("numRowsUpdated:" + numRowsUpdated);
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/AnomalyTaskSpecDAO.java]: use LOGGER

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/AnomalyTaskSpecDAO.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      return numRowsUpdated == 1;
+    } catch (HibernateException exception) {
+      exception.printStackTrace();
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/AnomalyTaskSpecDAO.java]: LOGGER

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/TaskDriver.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+          }
+          return null;
+        }
+      };
+      callables.add(callable);
+    }
+    for (Callable<Void> callable : callables) {
+      taskExecutorService.submit(callable);
+    }
+    System.out.println(Thread.currentThread().getId() + " : Started task driver");
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/TaskDriver.java]: LOG

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/TaskDriver.java [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+            LOG.info(Thread.currentThread().getId() + " : Finding next task to execute for threadId:{}",
+                Thread.currentThread().getId());
+
+            // select a task to execute, and update it to RUNNING
+            AnomalyTaskSpec anomalyTaskSpec = selectAndUpdate();
+            LOG.info(Thread.currentThread().getId() + " : Executing task: {} {}", anomalyTaskSpec.getTaskId(),
+                anomalyTaskSpec.getTaskInfo());
+
+            // execute the selected task
+            TaskRunner taskRunner = new TaskRunner(anomalyFunctionFactory);
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/TaskDriver.java]: extract outside loop

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeTask.java [hunk 11]
```diff
@@ -0,0 +1,5 @@
+package com.linkedin.thirdeye.anomaly;
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeTask.java]: remove if duplicate of Task

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeJobInfo.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package com.linkedin.thirdeye.anomaly;
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeJobInfo.java]: remove if duplicate of JobInfo

File: thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/anomaly/TestThirdEyeAnomalyApplication.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
+package com.linkedin.thirdeye.anomaly;
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/anomaly/TestThirdEyeAnomalyApplication.java]: remove

File: thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/anomaly/TesterClass.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.linkedin.thirdeye.anomaly.JobRunner.JobStatus;
+import com.linkedin.thirdeye.detector.api.AnomalyJobSpec;
+import com.linkedin.thirdeye.detector.api.AnomalyTaskSpec;
+import com.linkedin.thirdeye.detector.db.AnomalyFunctionSpecDAO;
+import com.linkedin.thirdeye.detector.db.AnomalyJobSpecDAO;
+import com.linkedin.thirdeye.detector.db.AnomalyTaskSpecDAO;
+
+
+public class TesterClass {
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/anomaly/TesterClass.java]: rename appropriately

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/JobScheduler.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    }
+  }
+
+  public void start() throws SchedulerException{
+    quartzScheduler.start();
+
+    //read the anomaly function specs
+    List<AnomalyFunctionSpec>  functionSpecs = readAnomalyFunctionSpecs();
+    for (AnomalyFunctionSpec anomalyFunctionSpec : functionSpecs) {
+      thirdEyeJobContext = new ThirdEyeJobContext();
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/JobScheduler.java]: check if it's active first

**Comment by jiateoh** [comment]
LGTM

**Comment by npawar** [comment]
Testing done: Started scheduler and multiple workers and checked for state transition of new tasks from WAITING to RUNNING to COMPLETED.

**Pull request merged**